File: solutions/javascript/LongestValidParentheses/LongestValidParentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    let arr = [-1], max = 0;
    for (let i = 0; i < s.length; i++) {
        let t = arr[arr.length - 1];
        if (t != -1 && s[i] == ')' && s[t] == '(') {
            arr.pop();
            max = Math.max(max, i - arr[arr.length - 1]);
        } else {
            arr.push(i);
        }
    }
    return max;
};

File: solutions/java/LongestValidParentheses/LongestValidParentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        Stack<Integer> stack = new Stack<>();
        stack.push(-1);
        int max = 0;
        for (int i = 0; i < s.length(); i++) {
            int t = stack.peek();
            if (t != -1 && s.charAt(i) == ')' && s.charAt(t) == '(') {
                stack.pop();
                max = Math.max(max, i - stack.peek());
            } else {
                stack.push(i);
            }
        }
        return max;
    }
}